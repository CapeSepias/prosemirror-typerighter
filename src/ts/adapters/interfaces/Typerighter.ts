export interface TypeRighterResponse {
  input: string;
  results: TypeRighterMatch[];
}

export interface TypeRighterMatch {
  fromPos: number;
  toPos: number;
  message: string;
  shortMessage: string;
  rule: TypeRighterRule;
  suggestedReplacements: string[];
}

export interface TypeRighterReplacement {
  value: string;
}

export interface TypeRighterType {
  typeName: string;
}

export interface TypeRighterRule {
  category: TypeRighterCategory;
  description: string;
  id: string;
  issueType: string;
}

export interface TypeRighterCategory {
  id: string;
  name: string;
}
