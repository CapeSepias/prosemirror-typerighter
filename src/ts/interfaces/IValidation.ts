export interface IRange {
  from: number;
  to: number;
}

export interface IValidationInput {
  id: string;
  str: string;
  from: number;
  to: number;
}

export type IValidationOutput = IValidationInput & {
  annotation: string;
  suggestions?: string[];
  type: string;
};

export interface IValidationError {
  validationInput: IValidationInput;
  message: string;
}

export interface IValidationResponse {
  validationOutputs: IValidationOutput[];
  validationInput: IValidationInput;
}

export type IValidationLibrary = Array<
  Array<{
    regExp: RegExp;
    annotation: string;
    operation: "ANNOTATE" | "REPLACE";
    type: string;
  }>
>;

export const Operations: {
  [key: string]: "ANNOTATE" | "REPLACE";
} = {
  ANNOTATE: "ANNOTATE",
  REPLACE: "REPLACE"
};
