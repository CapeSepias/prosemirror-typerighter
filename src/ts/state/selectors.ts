import { IMatches, IBlockResult as IMatches } from "../interfaces/IValidation";
import {
  IPluginState,
  IBlockQueryInFlight,
  IBlockQueriesInFlightState
} from "./reducer";

export const selectBlockQueriesInFlight = <TValidationMeta extends IMatches>(
  state: IPluginState<TValidationMeta>
) => {
  return state.blockQueriesInFlight.validations;
};

export const selectBlockMatchesByMatchId = <TValidationMeta extends IMatches>(
  state: IPluginState<TValidationMeta>,
  matchId: string
): IMatches | undefined =>
  state.currentValidations.find(validation => validation.matchId === matchId);

export const selectBlockQueriesInFlightForSet = <
  TValidationMeta extends IMatches
>(
  state: IPluginState<TValidationMeta>,
  requestId: string
): IBlockQueriesInFlightState | undefined => {
  return state.blockQueriesInFlight[requestId];
};

export const selectSingleBlockInFlightById = <
  TValidationMeta extends IMatches
>(
  state: IPluginState<TValidationMeta>,
  requestId: string,
  blockId: string
): IBlockQueryInFlight | undefined => {
  const validationInFlightState = selectBlockQueriesInFlightForSet(
    state,
    requestId
  );
  if (!validationInFlightState) {
    return;
  }
  return validationInFlightState.pendingBlocks.find(
    _ => _.block.id === blockId
  );
};

export const selectBlockQueriesInFlightById = <
  TValidationMeta extends IMatches
>(
  state: IPluginState<TValidationMeta>,
  requestId: string,
  blockIds: string[]
): IBlockQueryInFlight[] =>
  blockIds
    .map(blockId =>
      selectSingleBlockInFlightById(state, requestId, blockId)
    )
    .filter(_ => !!_) as IBlockQueryInFlight[];

export const selectAllBlockQueriesInFlight = <TValidationMeta extends IMatches>(
  state: IPluginState<TValidationMeta>
): IBlockQueryInFlight[] =>
  Object.values(state.blockQueriesInFlight).reduce(
    (acc, value) => acc.concat(value.pendingBlocks),
    [] as IBlockQueryInFlight[]
  );

type TSelectValidationInFlight = Array<
  IBlockQueriesInFlightState & {
    requestId: string;
  }
>;

export const selectNewBlockInFlight = <TValidationMeta extends IMatches>(
  oldState: IPluginState<TValidationMeta>,
  newState: IPluginState<TValidationMeta>
): TSelectValidationInFlight =>
  Object.keys(newState.blockQueriesInFlight).reduce(
    (acc, requestId) =>
      !oldState.blockQueriesInFlight[requestId]
        ? acc.concat({
            requestId,
            ...selectBlockQueriesInFlightForSet(newState, requestId)!
          })
        : acc,
    [] as TSelectValidationInFlight
  );

export const selectPercentRemaining = <TValidationMeta extends IMatches>(
  state: IPluginState<TValidationMeta>
) => {
  const [totalWork, totalRemainingWork] = Object.values(
    state.blockQueriesInFlight
  ).reduce(
    ([totalWorkAcc, remainingWorkAcc], queryState) => {
      const allWork =
        queryState.totalBlocks * queryState.categoryIds.length;
      const remainingWork = queryState.pendingBlocks.reduce(
        (acc, block) => acc + block.pendingCategoryIds.length,
        0
      );
      return [totalWorkAcc + allWork, remainingWorkAcc + remainingWork];
    },
    [0, 0]
  );
  return totalRemainingWork ? (totalRemainingWork / totalWork) * 100 : 0;
};

export const selectSuggestionAndRange = <TValidationMeta extends IMatches>(
  state: IPluginState<TValidationMeta>,
  matchId: string,
  suggestionIndex: number
) => {
  const output = selectBlockMatchesByMatchId(state, matchId);
  if (!output) {
    return null;
  }
  const suggestion = output.suggestions && output.suggestions[suggestionIndex];
  if (!suggestion) {
    return null;
  }
  return {
    from: output.from,
    to: output.to,
    suggestion
  };
};

export const selectAllAutoFixableValidations = <
  TMatches extends IMatches
>(
  state: IPluginState<TMatches>
): TMatches[] =>
  state.currentValidations.filter(_ => _.autoApplyFirstSuggestion);
