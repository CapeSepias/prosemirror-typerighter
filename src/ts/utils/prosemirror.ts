import { MarkSpec, Node, DOMOutputSpec } from 'prosemirror-model';
import { Transaction } from 'prosemirror-state';
import { ReplaceAroundStep, ReplaceStep } from 'prosemirror-transform';
import { IValidationInput } from '../interfaces/IValidation';

/**
 * Get a single string of text, and an array of position mappings,
 * from a Prosemirror document. The mappings can be used to map an
 * index in the text back to a position in the document.
 */
export const getTextMaps = (doc: Node): IValidationInput[] =>
  (doc instanceof Node ? findTextNodes(doc) : [doc]).reduce(
    (
      acc: { positionMap: IValidationInput[]; length: number },
      textNodeWrapper,
      index,
      textNodes
    ) => {
      const previousMap = acc.positionMap[acc.positionMap.length - 1];
      const str = textNodeWrapper.node.text || "";
      const previousNodeWrapper = textNodes[index - 1];
      const sharesParentWithPreviousNode =
        previousNodeWrapper &&
        textNodeWrapper.parent === previousNodeWrapper.parent;
      if (sharesParentWithPreviousNode) {
        // If this node shares a parent with the previous, add its text and
        // mapping to the last node's position map. In this way, contiguous
        // text nodes are treated as single lines of text for validation.
        const previousPositionMaps = acc.positionMap.slice(
          0,
          acc.positionMap.length - 1
        );
        const currentText = (previousMap ? previousMap.str : "") + str;
        return {
          length: acc.length + str.length,
          positionMap: previousPositionMaps.concat({
            str: currentText,
            from: previousMap.from,
            to: previousMap.from + currentText.length
          })
        };
      }
      // Add a new position map.
      return {
        length: acc.length + str.length,
        positionMap: acc.positionMap.concat({
          str,
          from: textNodeWrapper.pos,
          to: textNodeWrapper.pos + str.length
        })
      };
    },
    {
      positionMap: [],
      length: 0
    }
  ).positionMap;

export const MarkTypes = {
  legal: "legal",
  warn: "warn"
};

/**
 * Create a validation mark spec for the given mark name.
 */
const createValidationMark = (markName: string) => ({
  attrs: {},
  inclusive: false,
  parseDOM: [
    {
      tag: `span.${markName}`,
      getAttrs: () => ({})
    }
  ],
  toDOM: (): DOMOutputSpec => [`span.${markName}`]
});

export const validationMarks: {[name: string]: MarkSpec} = Object.keys(MarkTypes).reduce(
  (acc, markName: string) => {
    return {
      ...acc,
      [markName]: createValidationMark(markName)
    };
  },
  {} as { [markName: string]: MarkSpec }
);

/**
 * Flatten a node and its children into a single array of objects, containing
 * the node and the node's position in the document.
 */
export const flatten = (node: Node, descend = true) => {
  if (!node) {
    throw new Error('Invalid "node" parameter');
  }
  const result: Array<{ node: Node; parent: Node; pos: number }> = [];
  node.descendants((child, pos, parent) => {
    result.push({ node: child, parent, pos });
    if (!descend) {
      return false;
    }
  });
  return result;
};

/**
 * Find all children in a node that satisfy the given predicate.
 */
export const findChildren = (
  node: Node,
  predicate: (node: Node) => boolean,
  descend: boolean
): Array<{ node: Node; parent: Node; pos: number }> => {
  if (!node) {
    throw new Error('Invalid "node" parameter');
  } else if (!predicate) {
    throw new Error('Invalid "predicate" parameter');
  }
  return flatten(node, descend).filter(child => predicate(child.node));
};

/**
 * Find any text nodes in the given node.
 */
export const findTextNodes = (
  node: Node,
  descend: boolean = true
): Array<{ node: Node; parent: Node; pos: number }> => {
  return findChildren(node, child => child.isText, descend);
};

/**
 * Get all of the ranges of any replace steps in the given transaction.
 */
export const getReplaceStepRangesFromTransaction = (tr: Transaction) =>
  getReplaceTransactions(tr).map((step: ReplaceStep | ReplaceAroundStep) => ({
    from: (step as any).from,
    to: (step as any).to
  }));

/**
 * Get all of the ranges of any replace steps in the given transaction.
 */
export const getReplaceTransactions = (tr: Transaction) =>
  tr.steps.filter(
    step => step instanceof ReplaceStep || step instanceof ReplaceAroundStep
  );