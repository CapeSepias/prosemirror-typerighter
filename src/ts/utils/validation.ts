import { Transaction } from "prosemirror-state";
import { IRange } from "../interfaces/IValidation";

export const createValidationInput = (tr: Transaction, range: IRange) => {
  const inputString = tr.doc.textBetween(range.from, range.to);
  return {
    inputString,
    ...range,
    validationId: createValidationId(tr.time, range.from, range.to)
  };
};

export const createValidationId = (time: number, from: number, to: number) =>
  `${time}-from:${from}-to:${to}`;

export const createMatchId = (
  time: number,
  from: number,
  to: number,
  index: number = 0
) => `${createValidationId(time, from, to)}--match-${index}`;
