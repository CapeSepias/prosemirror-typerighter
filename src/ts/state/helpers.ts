import { DecorationSet } from "prosemirror-view";
import { Mapping } from "prosemirror-transform";
import { EditorState, Transaction } from "prosemirror-state";
import { Node } from "prosemirror-model";
import {
  IPluginState,
  IIgnoreMatchPredicate,
  includeAllMatches
} from "./reducer";
import {
  createDebugDecorationFromRange,
  createDecorationsForMatch,
  createDecorationsForMatches,
  removeDecorationsFromRanges
} from "../utils/decoration";
import { TFilterMatches } from "../utils/plugin";
import { getDirtiedRangesFromTransaction } from "../utils/prosemirror";
import { findOverlappingRangeIndex, mapAndMergeRanges, mapRanges } from "../utils/range";
import { IMatch, IRange } from "../interfaces/IMatch";

export const addMatchesToState = <TPluginState extends IPluginState>(
  state: TPluginState,
  doc: any,
  matches: Array<TPluginState["currentMatches"][number]>,
  ignoreMatch: IIgnoreMatchPredicate = includeAllMatches
) => {
  const matchesToApply = matches.filter(match => !ignoreMatch(match) && isMatchValid(match));
  const decorations = matchesToApply.reduce(
    (set, output) => set.add(doc, createDecorationsForMatch(output)),
    DecorationSet.empty
  );
  return {
    ...state,
    currentMatches: matchesToApply,
    decorations
  };
};

export const isMatchValid = (match: IMatch) => match.from < match.to;

/**
 * Is the current filter state stale, given the incoming state?
 */
export const isFilterStateStale = <TPluginState extends IPluginState>(
  oldState: TPluginState,
  newState: TPluginState,
  filterMatches?: TFilterMatches<TPluginState["filterState"]>
): filterMatches is TFilterMatches<TPluginState["filterState"]> => {
  const matchesChanged = oldState.currentMatches !== newState.currentMatches;
  const filterStateChanged = oldState.filterState !== newState.filterState;
  const noFilterApplied = !oldState.filterState && !newState.filterState;

  return (
    !!filterMatches &&
    (filterStateChanged || (matchesChanged && !noFilterApplied))
  );
};


/**
 * Get a new plugin state from the incoming transaction.
 *
 * We need to respond to each transaction in our reducer, whether or not there's
 * an action present, in order to maintain mappings and respond to user input.
 */
export const getNewStateFromTransaction = <TPluginState extends IPluginState>(
  tr: Transaction,
  pluginState: TPluginState,
  oldState: EditorState
): TPluginState => {
  if (!tr.docChanged) {
    return pluginState;
  }

  const newDirtiedRanges = getDirtiedRangesFromTransaction(oldState.doc, tr);
  const newPluginState = newDirtiedRanges.length
    ? applyNewDirtyRanges(tr, pluginState, newDirtiedRanges)
    : pluginState;

  const mappedRequestsInFlight = Object.entries(
    newPluginState.requestsInFlight
  ).reduce((acc, [requestId, requestsInFlight]) => {
    // We create a new mapping here to preserve state immutability, as
    // appendMapping mutates an existing mapping.
    const mapping = new Mapping();
    mapping.appendMapping(requestsInFlight.mapping);
    mapping.appendMapping(tr.mapping);
    return {
      ...acc,
      [requestId]: {
        ...requestsInFlight,
        mapping
      }
    };
  }, {});

  return {
    ...newPluginState,
    decorations: newPluginState.decorations.map(tr.mapping, tr.doc),
    dirtiedRanges: mapAndMergeRanges(newPluginState.dirtiedRanges, tr.mapping),
    currentMatches: mapRanges(newPluginState.currentMatches, tr.mapping),
    requestsInFlight: mappedRequestsInFlight,
    docChangedSinceCheck: true
  };
};


const applyNewDirtyRanges = <TPluginState extends IPluginState>(
  tr: Transaction,
  state: TPluginState,
  dirtiedRanges: IRange[]
): TPluginState => {
  // Map our dirtied ranges through the current transaction, and append any new ranges it has dirtied.
  let newDecorations = state.config.debug
    ? state.decorations.add(
        tr.doc,
        dirtiedRanges.map(range => createDebugDecorationFromRange(range))
      )
    : state.decorations;

  // Remove any matches and associated decorations touched by the dirtied ranges from the doc
  newDecorations = removeDecorationsFromRanges(newDecorations, dirtiedRanges);
  const currentMatches = state.currentMatches.filter(
    match => findOverlappingRangeIndex(match, dirtiedRanges) === -1
      && isMatchValid(match)
  );

  return {
    ...state,
    currentMatches,
    decorations: newDecorations,
    // We only care about storing dirtied ranges if we're validating
    // in response to user edits.
    requestPending: state.config.requestMatchesOnDocModified ? true : false,
    dirtiedRanges: state.config.requestMatchesOnDocModified
      ? state.dirtiedRanges.concat(dirtiedRanges)
      : []
  };
};

export const deriveFilteredDecorations = <TPluginState extends IPluginState>(
  doc: Node,
  newState: TPluginState,
  filterMatches: TFilterMatches<
    TPluginState["filterState"],
    TPluginState["currentMatches"][number]
  >
): TPluginState => {
  const filteredMatches = filterMatches(
    newState.filterState,
    newState.currentMatches
  );
  const filteredMatchIds = filteredMatches.map(_ => _.matchId);

  const matchIdsWithDecorations = newState.decorations
    .find()
    .map(_ => _.spec.id);

  const matchesWithoutDecorations = filteredMatches.filter(
    match => !matchIdsWithDecorations.includes(match.matchId)
  );

  const decorationsToAdd = createDecorationsForMatches(
    matchesWithoutDecorations,
    newState.config.matchColours
  );

  const decorationsToRemove = newState.decorations.find(
    undefined,
    undefined,
    spec => !filteredMatchIds.includes(spec.id)
  );

  const decorations = newState.decorations
    .remove(decorationsToRemove)
    .add(doc, decorationsToAdd);

  return {
    ...newState,
    filteredMatches,
    decorations
  };
};
