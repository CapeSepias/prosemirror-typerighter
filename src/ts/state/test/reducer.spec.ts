import { Transaction } from "prosemirror-state";
import { DecorationSet } from "prosemirror-view";
import {
  selectMatch,
  setDebugState,
  applyNewDirtiedRanges,
  validationRequestForDocument,
  validationRequestError,
  validationRequestForDirtyRanges,
  validationRequestSuccess,
  newHoverIdReceived
} from "../actions";
import { selectBlockQueriesInFlightForSet } from "../selectors";
import { createValidationPluginReducer, IPluginState } from "../reducer";
import {
  createDebugDecorationFromRange,
  getNewDecorationsForCurrentValidations,
  createDecorationForValidationRange
} from "../../utils/decoration";
import { expandRangesToParentBlockNode } from "../../utils/range";
import { createDoc, p } from "../../test/helpers/prosemirror";
import { IMatches } from "../../interfaces/IValidation";
import {
  createValidationResponse,
  createBlockQuery,
  exampleCategoryIds,
  createBlockQueriesInFlight,
  exampleValidationSetId,
  createInitialData,
  defaultDoc,
  addOutputsToState,
  createBlockMatches
} from "../../test/helpers/fixtures";
import { createValidationId } from "../../utils/validation";

const reducer = createValidationPluginReducer(expandRangesToParentBlockNode);

describe("Action handlers", () => {
  describe("No action", () => {
    it("should just return the state", () => {
      const { state, tr } = createInitialData();
      expect(reducer(tr, state)).toEqual(state);
    });
  });
  describe("Unknown action", () => {
    const { state, tr } = createInitialData();
    expect(reducer(tr, state, { type: "UNKNOWN_ACTION" } as any)).toEqual(
      state
    );
  });
  describe("validationRequestForDocument", () => {
    it("should apply dirty ranges for the entire doc", () => {
      const { state, tr } = createInitialData();
      expect(
        reducer(
          tr,
          state,
          validationRequestForDocument(
            exampleValidationSetId,
            exampleCategoryIds
          )
        )
      ).toMatchObject({
        blockQueriesInFlight: createBlockQueriesInFlight(
          exampleValidationSetId,
          [
            {
              from: 1,
              text: "Example text to validate",
              to: 26,
              id: "0-from:1-to:26"
            }
          ]
        )
      });
    });
  });
  describe("validationRequestForDirtyRanges", () => {
    it("should remove the pending status and any dirtied ranges, and mark the validation as in flight", () => {
      const { state, tr } = createInitialData();
      expect(
        reducer(
          tr,
          {
            ...state,
            debug: true,
            dirtiedRanges: [{ from: 5, to: 10 }],
            validationPending: true
          },
          validationRequestForDirtyRanges(
            exampleValidationSetId,
            exampleCategoryIds
          )
        )
      ).toEqual({
        ...state,
        debug: true,
        dirtiedRanges: [],
        decorations: new DecorationSet().add(tr.doc, [
          createDebugDecorationFromRange({ from: 1, to: 25 }, false)
        ]),
        validationPending: false,
        blockQueriesInFlight: createBlockQueriesInFlight(
          exampleValidationSetId,
          [
            {
              text: "Example text to validate",
              from: 1,
              to: 25,
              id: "0-from:1-to:25"
            }
          ]
        )
      });
    });
    it("should remove debug decorations, if any", () => {
      const { state, tr } = createInitialData();
      const newState = reducer(
        tr,
        {
          ...state,
          debug: true,
          dirtiedRanges: [{ from: 5, to: 10 }],
          decorations: new DecorationSet().add(tr.doc, [
            createDebugDecorationFromRange({ from: 1, to: 3 })
          ]),
          validationPending: true
        },
        validationRequestForDirtyRanges("id", exampleCategoryIds)
      );
      expect(newState.decorations).toEqual(
        new DecorationSet().add(tr.doc, [
          createDebugDecorationFromRange({ from: 1, to: 25 }, false)
        ])
      );
    });
    it("should add a total to the validations in flight", () => {
      const doc = createDoc(
        p("Example text to validate"),
        p("More text to validate")
      );
      const { state, tr } = createInitialData(doc);
      const newState = reducer(
        tr,
        {
          ...state,
          debug: true,
          dirtiedRanges: [{ from: 5, to: 10 }, { from: 28, to: 35 }],
          decorations: new DecorationSet(),
          validationPending: true
        },
        validationRequestForDirtyRanges("id", exampleCategoryIds)
      );
      expect(selectBlockQueriesInFlightForSet(newState, "id")!.total).toEqual(
        2
      );
    });
  });
  describe("validationRequestSuccess", () => {
    it("shouldn't do anything if there's nothing in the response and nothing to clean up", () => {
      const { state, tr } = createInitialData();
      expect(
        reducer(
          tr,
          state,
          validationRequestSuccess({
            blocks: [{ from: 1, to: 1, text: "hai", id: "1" }],
            matches: [],
            categoryIds: ["some-cat"],
            validationSetId: exampleValidationSetId
          })
        )
      ).toEqual(state);
    });
    it("shouldn't do anything if there are no categories", () => {
      const { state, tr } = createInitialData();
      expect(
        reducer(
          tr,
          state,
          validationRequestSuccess({
            blocks: [{ from: 1, to: 1, text: "hai", id: "1" }],
            matches: [],
            categoryIds: [],
            validationSetId: exampleValidationSetId
          })
        )
      ).toEqual(state);
    });
    it("should add incoming validations to the state", () => {
      const { state, tr } = createInitialData();
      let localState = reducer(
        tr,
        state,
        applyNewDirtiedRanges([{ from: 1, to: 3 }])
      );
      localState = reducer(
        tr,
        localState,
        validationRequestForDirtyRanges(
          exampleValidationSetId,
          exampleCategoryIds
        )
      );
      expect(
        reducer(
          tr,
          localState,
          validationRequestSuccess(createValidationResponse(1, 25))
        ).currentValidations
      ).toMatchObject([createBlockMatches(1, 4)]);
    });
    it("should create decorations for the incoming validations", () => {
      const { state, tr } = createInitialData();
      expect(
        reducer(
          tr,
          state,
          validationRequestSuccess(createValidationResponse(5, 10))
        )
      ).toMatchSnapshot();
    });
    describe("superceded matches", () => {
      const { state: initialState, tr } = createInitialData();
      const firstBlock = createBlockQuery(0, 15, "Example text to validate");
      const secondBlock = createBlockQuery(16, 37, "Another block of text");
      const blocks = [firstBlock, secondBlock];
      const category = {
        id: "this-category-should-remain",
        colour: "purple",
        name:
          "This category should remain untouched -- it's not included in the categories for the incoming matches"
      };
      const validationOutput1 = createValidationResponse(0, 15, 1, 7);
      const validationOutput2 = createValidationResponse(
        0,
        15,
        9,
        13,
        category
      );
      const validationOutput3 = createValidationResponse(16, 37, 17, 25); // Some other output for another block
      const blockQueriesInFlight = createBlockQueriesInFlight(
        exampleValidationSetId,
        blocks,
        [...validationOutput1.categoryIds, ...validationOutput2.categoryIds]
      );

      const state: IPluginState = addOutputsToState(
        {
          ...initialState,
          blockQueriesInFlight
        },
        tr.doc,
        [
          validationOutput1.matches[0],
          validationOutput2.matches[0],
          validationOutput3.matches[0]
        ]
      );

      it("should remove previous validations that match the block and category of the incoming match", () => {
        const newState = reducer(
          tr,
          state,
          validationRequestSuccess({
            blocks: [firstBlock],
            categoryIds: ["1"],
            matches: [],
            validationSetId: exampleValidationSetId
          })
        );

        expect(newState.currentValidations).toEqual([
          validationOutput2.matches[0],
          validationOutput3.matches[0]
        ]);
      });
      it("should remove previous decorations that match block and category of the incoming match", () => {
        const newState = reducer(
          tr,
          state,
          validationRequestSuccess({
            blocks: [firstBlock],
            categoryIds: ["1"],
            matches: [],
            validationSetId: exampleValidationSetId
          })
        );

        expect(newState.decorations).toEqual(
          new DecorationSet().add(tr.doc, [
            ...createDecorationForValidationRange(validationOutput2.matches[0]),
            ...createDecorationForValidationRange(validationOutput3.matches[0])
          ])
        );
      });

      it("should remove validated categories from the remaining categories list for in-flight blocks", () => {
        const newState = reducer(
          tr,
          state,
          validationRequestSuccess({
            blocks: [firstBlock],
            categoryIds: ["1"],
            matches: [],
            validationSetId: exampleValidationSetId
          })
        );

        expect(newState.blockQueriesInFlight).toEqual({
          "set-id": {
            current: [
              {
                allCategoryIds: ["1", "this-category-should-remain"],
                blockQuery: {
                  from: 0,
                  id: "0-from:0-to:15",
                  text: "Example text to validate",
                  to: 15
                },
                mapping: { from: 0, maps: [], mirror: undefined, to: 0 },
                remainingCategoryIds: ["this-category-should-remain"]
              },
              {
                allCategoryIds: ["1", "this-category-should-remain"],
                blockQuery: {
                  from: 16,
                  id: "0-from:16-to:37",
                  text: "Another block of text",
                  to: 37
                },
                mapping: { from: 0, maps: [], mirror: undefined, to: 0 },
                remainingCategoryIds: ["1", "this-category-should-remain"]
              }
            ],
            total: 2
          }
        });
      });
    });
    it("should not apply validations if the ranges they apply to have since been dirtied", () => {
      const { state, tr } = createInitialData(defaultDoc, 1337);
      let localState = reducer(
        tr,
        state,
        applyNewDirtiedRanges([{ from: 1, to: 3 }])
      );
      localState = reducer(
        tr,
        localState,
        validationRequestForDirtyRanges("id", exampleCategoryIds)
      );
      localState = reducer(
        tr,
        localState,
        applyNewDirtiedRanges([{ from: 1, to: 3 }])
      );
      expect(
        reducer(
          tr,
          localState,
          validationRequestSuccess(createValidationResponse(1, 3))
        )
      ).toEqual({
        ...localState,
        dirtiedRanges: [{ from: 1, to: 3 }],
        currentValidations: [],
        validationPending: true
      });
    });
  });
  describe("validationRequestError", () => {
    it("Should re-add the in-flight validation ranges as dirty ranges, and remove the inflight validation", () => {
      const { state: initialState, tr } = createInitialData();
      const state = {
        ...initialState,
        blockQueriesInFlight: createBlockQueriesInFlight(
          exampleValidationSetId,
          [createBlockQuery(1, 25, "Example text to validate")]
        )
      };
      const newState = reducer(
        tr,
        state,
        validationRequestError({
          validationSetId: exampleValidationSetId,
          validationId: createValidationId(0, 1, 25),
          message: "Too many requests"
        })
      );
      expect(newState).toMatchObject({
        blockQueriesInFlight: {},
        dirtiedRanges: [
          {
            from: 1,
            to: 25
          }
        ],
        decorations: new DecorationSet(),
        error: "Too many requests"
      });
    });
  });
  describe("newHoverIdReceived", () => {
    it("should update the hover id", () => {
      const { state } = createInitialData();
      expect(
        reducer(
          new Transaction(createDoc),
          state,
          newHoverIdReceived("exampleHoverId", undefined)
        )
      ).toEqual({
        ...state,
        hoverId: "exampleHoverId",
        hoverInfo: undefined
      });
    });
    it("should add hover decorations", () => {
      const { state, tr } = createInitialData();
      const output: IMatches = {
        matchId: "match-id",
        from: 0,
        to: 5,
        annotation: "Annotation",
        category: {
          id: "1",
          name: "cat",
          colour: "eeeeee"
        }
      };
      const localState = {
        ...state,
        currentValidations: [output],
        decorations: new DecorationSet().add(
          tr.doc,
          createDecorationForValidationRange(output, false, true)
        )
      };
      expect(reducer(tr, localState, newHoverIdReceived("match-id"))).toEqual({
        ...localState,
        decorations: new DecorationSet().add(
          tr.doc,
          createDecorationForValidationRange(output, true, true)
        ),
        hoverId: "match-id",
        hoverInfo: undefined
      });
    });
    it("should remove hover decorations", () => {
      const { state, tr } = createInitialData();
      const output: IMatches = {
        matchId: "match-id",
        from: 0,
        to: 5,
        annotation: "Annotation",
        category: {
          id: "1",
          name: "cat",
          colour: "eeeeee"
        }
      };
      const localState = {
        ...state,
        decorations: new DecorationSet().add(tr.doc, [
          ...createDecorationForValidationRange(output, true, true)
        ]),
        currentValidations: [output],
        hoverId: "match-id",
        hoverInfo: undefined
      };
      expect(
        reducer(tr, localState, newHoverIdReceived(undefined, undefined))
      ).toEqual({
        ...localState,
        decorations: new DecorationSet().add(tr.doc, [
          ...createDecorationForValidationRange(output, false, true)
        ]),
        hoverId: undefined,
        hoverInfo: undefined
      });
    });
  });
  describe("handleNewDirtyRanges", () => {
    it("should remove any decorations and validations that touch the passed ranges", () => {
      const { state } = createInitialData();
      const currentValidations: IMatches[] = [
        {
          matchId: "match-id",
          from: 1,
          to: 7,
          annotation: "Annotation",
          category: {
            id: "1",
            name: "cat",
            colour: "eeeeee"
          }
        }
      ];
      const stateWithCurrentValidationsAndDecorations = {
        ...state,
        currentValidations,
        decorations: getNewDecorationsForCurrentValidations(
          currentValidations,
          state.decorations,
          defaultDoc
        )
      };
      expect(
        reducer(
          new Transaction(defaultDoc),
          stateWithCurrentValidationsAndDecorations,
          applyNewDirtiedRanges([{ from: 1, to: 2 }])
        )
      ).toEqual({
        ...state,
        validationPending: true,
        dirtiedRanges: [{ from: 1, to: 2 }]
      });
    });
  });
  describe("selectValidation", () => {
    it("should apply the selected validation id", () => {
      const { state } = createInitialData();
      const otherState = {
        ...state,
        currentValidations: [
          {
            matchId: "match-id",
            text: "example",
            from: 1,
            to: 1,
            annotation: "example",
            suggestions: [],
            category: {
              id: "1",
              name: "cat",
              colour: "eeeeee"
            },
            id: "exampleId"
          }
        ]
      };
      expect(
        reducer(
          new Transaction(createDoc),
          otherState,
          selectMatch("exampleId")
        )
      ).toEqual({
        ...otherState,
        selectedMatch: "exampleId"
      });
    });
  });
  describe("setDebug", () => {
    it("should set the debug state", () => {
      const { state } = createInitialData();
      expect(
        reducer(new Transaction(createDoc), state, setDebugState(true))
      ).toEqual({ ...state, debug: true });
    });
  });
});
