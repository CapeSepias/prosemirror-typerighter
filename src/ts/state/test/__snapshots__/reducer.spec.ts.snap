// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action handlers requestMatchesSuccess should create decorations for the incoming matches 1`] = `
Object {
  "currentMatches": Array [],
  "currentThrottle": 100,
  "debug": false,
  "decorations": DecorationSet {
    "children": Array [],
    "local": Array [],
  },
  "dirtiedRanges": Array [],
  "error": undefined,
  "hoverId": undefined,
  "hoverInfo": undefined,
  "initialThrottle": 100,
  "maxThrottle": 1000,
  "requestMatchesOnDocModified": true,
  "requestPending": false,
  "requestsInFlight": Object {},
  "selectedMatch": undefined,
  "trHistory": Array [
    Transaction {
      "curSelection": undefined,
      "curSelectionFor": 0,
      "doc": Object {
        "content": Array [
          Object {
            "content": Array [
              Object {
                "text": "Example text to check",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "doc",
      },
      "docs": Array [],
      "mapping": Mapping {
        "from": 0,
        "maps": Array [],
        "mirror": undefined,
        "to": 0,
      },
      "meta": Object {},
      "steps": Array [],
      "storedMarks": undefined,
      "time": 0,
      "updated": 0,
    },
  ],
}
`;
